src/cargo/core/resolver/mod.rs

```diff
@@ -986,21 +1077,51 @@ impl<'a> Context<'a> {
         // Record what list of features is active for this package.
         if !used_features.is_empty() {
             let pkgid = candidate.package_id();
-            self.resolve_features.entry(pkgid.clone())
-                .or_insert(HashSet::new())
-                .extend(used_features);
+
+            let mut set = self.resolve_features.entry(pkgid.clone())
+                              .or_insert_with(HashSet::new);
+            for feature in used_features {
+                if !set.contains(feature) {
+                    set.insert(feature.to_string());
+                }
+            }
         }
 
         Ok(ret)
     }
+
+    fn resolve_replacements(&self) -> HashMap<PackageId, PackageId> {
+        let mut replacements = HashMap::new();
+        let mut cur = &self.resolve_replacements;
+        while let Some(ref node) = cur.head {
+            let (k, v) = node.0.clone();
+            replacements.insert(k, v);
+            cur = &node.1;
+        }
+        return replacements
+    }
+
+    fn graph(&self) -> Graph<PackageId> {
+        let mut graph = Graph::new();
+        let mut cur = &self.resolve_graph;
+        while let Some(ref node) = cur.head {
+            match node.0 {
+                GraphNode::Add(ref p) => graph.add(p.clone(), &[]),
+                GraphNode::Link(ref a, ref b) => graph.link(a.clone(), b.clone()),
+            }
+            cur = &node.1;
+        }
+        return graph
+    }
 }
 
 fn check_cycles(resolve: &Resolve,
-                activations: &HashMap<(String, SourceId), Vec<Rc<Summary>>>)
+                activations: &HashMap<String, HashMap<SourceId, Vec<Summary>>>)
```

Comment:
We could typealias `activations`, if we indent to tweak this datastructure further.